### PR TITLE
Super fast search

### PR DESCRIPTION
**Pull request by ondrae**
This PR uses both a materialized view for our default search, then tsvector columns for when sections are selected.

#### ToDo:
- [x] Refactor search action in controller for simplicity /readability
- [x] Double check schema.sql if needed
- [x] Add a call to refresh the materialized view after our daily update.
- [x] Double check database size
- [x] Add documentation about search db approach, with links.
- [x] Non int handling in publication date, in controller.
- [x] Clean up html vs csv code in controller.
- [x] Add `FullSornSearch.refresh` to all factory create calls automatically.

### CODE REVIEW

File: app/jobs/refresh_search_job.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class RefreshSearchJob < ApplicationJob
```
Comment (ondrae) [inline app/jobs/refresh_search_job.rb:1]: This job is here just if we need it later, depending on how the daily update gets implemented.

**Review by lauraGgit — CHANGES_REQUESTED** [review]
I have a few comments about error handling. I also think that the readme needs some documentation about your fast search implementation and references to the gems used.

File: app/controllers/sorns_controller.rb [hunk 2]
```diff
@@ -1,16 +1,47 @@
...
+    @sorns = @sorns.joins(:agencies).where(agencies: {name: params[:agencies]})
+    # Matching on agencies could return duplicates, so get distinct
+    if @fields_to_search == Sorn::FIELDS
+       @sorns.get_distinct
+    else
+      @sorns.get_distinct_with_dynamic_search_rank
     end
+  end
 
+  def filter_on_publication_date
```
[the remark this answers is not in the document]
Comment (ondrae) [reply app/controllers/sorns_controller.rb:44]: Right, on it.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -10,7 +10,7 @@ jobs:
     runs-on: ubuntu-latest
     services:
       postgres:
-        image: postgres:11
+        image: postgres
```
[the remark this answers is not in the document]
Comment (ondrae) [reply .github/workflows/deploy.yml:13]: The postgres "generated" column type was introduced in postgres:12. Locally and on cloud.gov we are on 13. We could pin it to 13 if we want to be specific.

File: spec/requests/search_csv_request_spec.rb [hunk 4]
```diff
@@ -1,9 +1,12 @@
 require 'rails_helper'
 
 RSpec.describe "Search csv", type: :request do
-  let!(:sorn) { create :sorn, citation: "citation" }
 
-  before { get "/search.csv?search=#{search}&#{fields}&#{agency}" }
+  before do
+    create :sorn
+    FullSornSearch.refresh # refresh the materialized view
```
[the remark this answers is not in the document]
Comment (ondrae) [reply spec/requests/search_csv_request_spec.rb:7]: ooooh good idea, then we wouldn't have to remember to do it.

**Comment by ondrae** [comment]
I've added all @lauraGgit suggestions to the top comment to work through. Thanks for the review.

File: spec/requests/search_request_spec.rb [hunk 5]
```diff
@@ -1,12 +1,14 @@
 require "rails_helper"
 
 RSpec.describe "Search", type: :request do
-  let!(:sorn) { create :sorn, citation: "citation" }
+  let!(:sorn) { create :sorn }
```
Comment (lauraGgit) [inline spec/requests/search_request_spec.rb:4]: a neat example of the tests barely needing to change in the refactor!

**Review by lauraGgit — COMMENTED** [review]
LGTM.- I might take a look at the Kanmari issue that was popping up, but I wasn't able to reproduce locally.

**Review by lauraGgit — APPROVED** [review]

**Pull request merged**